lstchain/io/io.py

```diff
@@ -969,3 +974,70 @@ def write_calibration_data(writer, mon_index, mon_event, new_ped=False, new_ff=F
             table_name="telescope/monitoring/calibration",
             containers=[mon_index, mon_event.calibration]
         )
+
+
+def read_dl2_sub_run(t_filename, columns_to_read=None):
+    '''
+    Read a file with DL2 data
+
+    Parameters
+    ----------
+    t_filename: Input file name
+    columns_to_read: List of interesting columns, optional. If None, then all columns will be read
+
+    Returns
+    -------
+    Pandas dataframe with DL2 data
+    '''
+    if columns_to_read is not None:
+        return pd.read_hdf(t_filename, key=dl2_params_lstcam_key)[columns_to_read]
+    else:
+        return pd.read_hdf(t_filename, key=dl2_params_lstcam_key)
+
+
+def extract_observation_time(t_df):
+    '''
+    Calculate observation time
+
+    Parameters
+    ----------
+    pandas.DataFrame t_df: Recorded data
+
+    Returns
+    -------
+    Observation duration in seconds
+    '''
+    return pd.to_datetime(t_df.dragon_time.iat[len(t_df)-1], unit='s') -\
+           pd.to_datetime(t_df.dragon_time.iat[0], unit='s')
+
+
+def merge_dl2_runs(data_path, runs, columns_to_read=None, n_process=4):
+    """
+    Merge the run sequence in a single dataset and extract correct observation time based on first and last event timestamp in each file.
+
+    Parameters
+    ----------
+    data_path: Path to data location
+    runs: List of run numbers
+    n_process: Number of parallel read processes to use
+
+    Returns
+    -------
+    Pair (observation time, data)
+    """
+    from functools import partial
+
+    pool = Pool(n_process)
+    filelist = []
+    # Create a list of files with matching run numbers
+    for filename in sorted(os.listdir(data_path)):
+        if any(str(run) in filename for run in runs):
+            filelist.append(f'{data_path}/{filename}')
+
+    df_list = pool.map(partial(read_dl2_sub_run, columns_to_read=columns_to_read), filelist)
+
+    observation_times = pool.map(extract_observation_time, df_list)
+
+    observation_time = np.sum(observation_times).total_seconds()
```

Comment:
`np.sum(observation_times)` is a numpy array - it does not have a `total_seconds` attribute

Reply:
since `observation_times` is a 1D array, the return type is a scalar with its type inherited (datetime in this case), so this object has `total_seconds` attribute...

Reply:
Then it makes no sense to call np.sum in the first place. What exactly is observation_times and what the desired output?

Reply:
`observation_times ` is an array of subrun durations (for each subrun takes last-first timestamp). Its sum is a total duration of observation. The desired (and de-facto) output is the duration of observation in seconds.

Reply:
`pool.map` returns a `list` not an array.

Looking at the code of the extract function, that returns a `pandas.TimeDelta` object. 

So what you have is a python `list` of pandas `TimeDelta`.

You should just use the python sum and a list comprehension for that:


```python

In [15]: import numpy as np

In [16]: import pandas as pd

In [17]: obs_times = [pd.Timedelta(seconds=s) for s in np.random.uniform(1000, 2000, 100)]

In [18]: %timeit np.sum(obs_times).total_seconds()
493 µs ± 2.67 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [19]: %timeit sum([t.total_seconds() for t in obs_times])
```

Reply:
ah, thanks! Done.